File: benchmarks/server/driver/src/features.ts
import { SirannonClient } from './sirannon-client.ts'
import { maxOf, mean, percentile } from './stats.ts'

const PROBE_TABLE = 'cdc_probe'
const SERVER_POLL_INTERVAL_MS = 50
const SETUP_ATTEMPTS = 5
const SETUP_BACKOFF_MS = 250
const SETUP_TIMEOUT_MS = 10_000

function delay(ms: number): Promise<void> {
  return new Promise(resolve => {
    setTimeout(resolve, ms)
  })
}

export async function measureCdcLatency(
  baseUrl: string,
  databaseId: string,
  samples: number,
  warmupSamples: number,
): Promise<Record<string, unknown>> {
  const endpoint = `${baseUrl.replace(/\/+$/, '')}/db/${encodeURIComponent(databaseId)}`
  const post = async (path: string, body: unknown): Promise<void> => {
    const url = `${endpoint}${path}`
    const payload = JSON.stringify(body)
    let lastError: unknown
    for (let attempt = 1; attempt <= SETUP_ATTEMPTS; attempt++) {
      let response: Response
      try {
        response = await fetch(url, {
          method: 'POST',
          headers: { 'content-type': 'application/json' },
          body: payload,
          signal: AbortSignal.timeout(SETUP_TIMEOUT_MS),
        })
      } catch (err) {
        lastError = err
        if (attempt < SETUP_ATTEMPTS) {
          await delay(SETUP_BACKOFF_MS * attempt)
          continue
        }
        throw err
      }
      if (!response.ok) {
        throw new Error(`POST ${path} returned ${response.status}: ${await response.text()}`)
      }
      return
    }
    throw lastError
  }

  await post('/execute', { sql: `DROP TABLE IF EXISTS ${PROBE_TABLE}` })
  await post('/execute', { sql: `CREATE TABLE ${PROBE_TABLE} (id INTEGER PRIMARY KEY, marker INTEGER NOT NULL)` })

  const client = new SirannonClient(baseUrl)
  const db = client.database(databaseId)
  const latenciesMs: number[] = []

  let awaited: { marker: number; resolve: () => void } | null = null
  const subscription = await db.on(PROBE_TABLE).subscribe(event => {
    const row = (event.row ?? {}) as Record<string, unknown>
    if (awaited && (row.id === awaited.marker || row.marker === awaited.marker)) {
      const resolve = awaited.resolve
      awaited = null
      resolve()
    }
  })

  try {
    const total = warmupSamples + samples
    for (let index = 1; index <= total; index++) {
      const changeSeen = new Promise<void>(resolve => {
        awaited = { marker: index, resolve }
      })
      const sentAt = performance.now()
      await db.execute(`INSERT INTO ${PROBE_TABLE} (id, marker) VALUES (?, ?)`, [index, index])
      await changeSeen
      if (index > warmupSamples) {
        latenciesMs.push(performance.now() - sentAt)
      }
    }
  } finally {
    subscription.unsubscribe()
    client.close()
  }

  return {
    feature: 'cdc-latency',
    description:
      'Lag from a committed write to the change arriving at a subscriber over Sirannon\'s built-in ' +
      'WebSocket change feed. The server polls the change log on a fixed interval, so the floor is ' +
      'that interval.',
    samples: latenciesMs.length,
    server_poll_interval_ms: SERVER_POLL_INTERVAL_MS,
    latency_ms: {
      p50: percentile(latenciesMs, 0.5),
      p95: percentile(latenciesMs, 0.95),
      p99: percentile(latenciesMs, 0.99),
      max: maxOf(latenciesMs),
      mean: mean(latenciesMs),
    },
  }
}

File: benchmarks/server/src/sirannon_bench/features.py
"""Sirannon-only characterizations that PostgreSQL has no direct equivalent for.

These are framed as characterizations, never as a win over PostgreSQL, because the two systems
do not do the same work here. The change-feed measurement reports the lag from a committed write
to that change arriving at a subscriber over Sirannon's built-in WebSocket feed. PostgreSQL has
no built-in change feed, so the honest comparison would be against PostgreSQL plus an external
Debezium and Kafka pipeline, which is a stack comparison, not an engine one. The number here
describes Sirannon's feed on its own terms.
"""

from __future__ import annotations

import asyncio
import json

import httpx
import websockets

from .core.stats import mean, percentile

_PROBE_TABLE = "cdc_probe"
_SERVER_POLL_INTERVAL_MS = 50


def _ws_url(base_url: str, database_id: str) -> str:
    scheme = "wss" if base_url.startswith("https://") else "ws"
    authority = base_url.split("://", 1)[1].rstrip("/")
    return f"{scheme}://{authority}/db/{database_id}"


async def measure_cdc_latency(
    base_url: str,
    database_id: str,
    samples: int,
    warmup_samples: int,
) -> dict:
    endpoint = f"{base_url.rstrip('/')}/db/{database_id}"
    async with httpx.AsyncClient(timeout=httpx.Timeout(30.0)) as client:
        await client.post(f"{endpoint}/execute", json={"sql": f"DROP TABLE IF EXISTS {_PROBE_TABLE}"})
        await client.post(
            f"{endpoint}/execute",
            json={"sql": f"CREATE TABLE {_PROBE_TABLE} (id INTEGER PRIMARY KEY, marker INTEGER NOT NULL)"},
        )

        async with websockets.connect(_ws_url(base_url, database_id)) as socket:
            await socket.send(json.dumps({"type": "subscribe", "id": "probe", "table": _PROBE_TABLE}))
            await _await_message(socket, "subscribed")

            latencies_ms: list[float] = []
            loop = asyncio.get_running_loop()
            total = warmup_samples + samples
            for index in range(1, total + 1):
                sent_at = loop.time()
                await client.post(
                    f"{endpoint}/execute",
                    json={"sql": f"INSERT INTO {_PROBE_TABLE} (id, marker) VALUES (?, ?)", "params": [index, index]},
                )
                await _await_change_for(socket, index)
                if index > warmup_samples:
                    latencies_ms.append((loop.time() - sent_at) * 1000.0)

    return {
        "feature": "cdc-latency",
        "description": (
            "Lag from a committed write to the change arriving at a subscriber over Sirannon's "
            "built-in WebSocket change feed. The server polls the change log on a fixed interval, "
            "so the floor is that interval."
        ),
        "samples": len(latencies_ms),
        "server_poll_interval_ms": _SERVER_POLL_INTERVAL_MS,
        "latency_ms": {
            "p50": percentile(latencies_ms, 0.50),
            "p95": percentile(latencies_ms, 0.95),
            "p99": percentile(latencies_ms, 0.99),
            "max": max(latencies_ms) if latencies_ms else 0.0,
            "mean": mean(latencies_ms),
        },
    }


async def _await_message(socket, expected_type: str) -> dict:
    while True:
        message = json.loads(await socket.recv())
        if message.get("type") == "error":
            raise RuntimeError(f"change feed returned an error: {message.get('error')}")
        if message.get("type") == expected_type:
            return message


async def _await_change_for(socket, marker: int) -> dict:
    while True:
        message = await _await_message(socket, "change")
        row = message.get("event", {}).get("row", {})
        if row.get("id") == marker or row.get("marker") == marker:
            return message
